src/content-helper/editor-sidebar/smart-linking/component.tsx

```diff
@@ -179,130 +191,231 @@ export const SmartLinkingPanel = ( {
+	 * - `linked`: The number of times a link has been successfully applied for a specific link text.
+	 *
+	 * @since 3.14.1
+	 */
+	type LinkOccurrenceCounts = {
+		[key: string]: {
+			encountered: number;
+			linked: number;
+		};
+	};
+
 	/**
 	 * Applies the smart links to the selected block or the entire post content.
 	 *
 	 * @since 3.14.0
+	 * @since 3.14.1 Moved applyLinksToBlocks to a separate function.
 	 *
 	 * @param {LinkSuggestion[]} links The smart links to apply.
 	 */
 	const applySmartLinks = ( links: LinkSuggestion[] ): void => {
 		Telemetry.trackEvent( 'smart_linking_applied', {
-			is_full_content: fullContent,
+			is_full_content: isFullContent || ! selectedBlock,
 			selected_block: selectedBlock?.name ?? 'none',
 			links_count: links.length,
 			context,
 		} );
 
-		// Get the original content of the selected block or the entire post content.
-		let originalContent = '';
-		if ( selectedBlock && ! fullContent ) {
-			originalContent = selectedBlock.attributes.content;
+		let blocks;
+		if ( selectedBlock && ! isFullContent ) {
+			blocks = [ selectedBlock ];
 		} else {
-			originalContent = postContent;
+			blocks = allBlocks;
 		}
 
-		let newContent = originalContent; // Fallback to original content if no links are found.
-		for ( const link of links ) {
-			// Check if the content already contains the link, skip if so.
-			if ( originalContent.includes( link.title ) && originalContent.includes( link.href ) ) {
-				continue;
-			}
+		// An object to keep track of the number of times each link text has been found across all blocks.
+		const occurrenceCounts: LinkOccurrenceCounts = {};
+
+		// Apply the smart links to the content.
+		applyLinksToBlocks( blocks, links, occurrenceCounts );
 
-			// Escape the link text to convert regex special characters to literal characters.
-			link.text = escapeRegExp( link.text );
-
-			// Check if the amount of link.text occurrences in the newContent is bigger than the amount of
-			// link.text occurrences in the originalContent, if so, it means that we've introduced another
-			// occurrence of the link.text in the newContent, so we need to increase the offset.
-			const linkTextRegex = new RegExp( link.text, 'g' );
-			const newContentMatches = newContent.match( linkTextRegex );
-			const originalContentMatches = originalContent.match( linkTextRegex );
-			if ( ( newContentMatches && originalContentMatches ) &&
-				newContentMatches?.length > originalContentMatches?.length ) {
-				link.offset++;
+		// Update the content of each block.
+		blocks.forEach( ( block ) => {
+			updateBlockContent( block );
+		} );
+
+		createNotice( 'success', `${ links.length } smart links successfully applied.`, {
+			type: 'snackbar',
+			isDismissible: true,
+		} );
+	};
+
+	/**
+	 * Iterates through blocks of content to apply smart link suggestions.
+	 * This function parses the content of each block, looking for text nodes that match the provided link suggestions.
```

Comment:
```suggestion
	 * 
	 * This function parses the content of each block, looking for text nodes that match the provided link suggestions.
```
I think we should be leaving a blank line between the function summary and description. Also valid for other functions.